Add inventory monitor module

chimp1984 opened the pull request:
This adds a new module containing the new inventory monitor project. 
It is running since a few weeks under http://46.101.179.224/ and some of the seeds have updated to that branch already.
It is the implementation of https://github.com/bisq-network/projects/issues/45 
@jmacxx is working on a webapp which should finally replace the simple http server used atm.
It will still need fine tuning of the deviation alert/warning thresholds but it is already pretty useful and helps to detect seed node issues.
Most of the code is only relevant to the monitor project and seed nodes, so it carries very low risk for normal users.

chimp1984 committed:
Extract method for filling seed nodes from property file

chimp1984 committed:
Make config nullable and extract methods where config is used and apply default value in case config is not set.

chimp1984 committed:
Add inventory module

Simple monitor dumping json files with data result and request/response time to disk. Can be used by a simple web server to show state of seed nodes.

chimp1984 committed:
Move inventory package to core as we want to include other domain data as well like DAO state

Add more data to inventory map and change type of value to String.

chimp1984 committed:
Add signature to request to limit the feature to requests from trusted nodes (monitor)

chimp1984 committed:
Add readableFileSize

chimp1984 committed:
Add readSeedNodePropertyFile method

chimp1984 committed:
Add sparkjava dependency

chimp1984 committed:
Remove signature from GetInventoryRequest (as monitor is public it does not make sense to limit requests).
Add html output and webserver

chimp1984 committed:
Add shadow plugin

chimp1984 committed:
Make RequestInfo public

chimp1984 committed:
Add InventoryItem enum
Add average and color codes to html
Add maxConnections
Add DeviationSeverity enum
Add custom seed node file
Use InventoryItem as key in inventory map instead of string

chimp1984 committed:
Add nodes from mike
Various improvements...

chimp1984 committed:
Shut down connection after response received

Various improvements...

chimp1984 committed:
Add network node shutdown
Rearrange request info fields

chimp1984 committed:
Add more warn/alert checks

chimp1984 committed:
Add triggers for maxConnections and numConnections

chimp1984 committed:
Fix error handling
Add all seeds

chimp1984 committed:
Add peakNumConnections

chimp1984 committed:
Add numAllConnectionsLostEvents

chimp1984 commented:
@ripcurlx Please remove rule that empty catch with ignore as param name is causing warnings.

chimp1984 commented:
@ripcurlx Please remove rule that empty catch when using ignore is causing issues

ripcurlx commented:
> @ripcurlx Please remove rule that empty catch when using ignore is causing issues

Just removed the pattern.

chimp1984 committed:
Only log and increase numAllConnectionsLostEvents if we have not called shutDown
We also do not call listeners in that case as some listeners have some restart routines
on that event which is not what we want at shut down

chimp1984 committed:
Improve deviation model

chimp1984 committed:
Map shutdown call from sig int handler to userThread

chimp1984 committed:
Fix incorrect shutdown behaviour

Delay the boolean property setter as otherwise our listener might never
get triggered if property is set synchronously before listener registration.
Remove shutdown thread.
Cancel future in case tor is not created yet.

chimp1984 committed:
Remove torSetup.cleanupTorFiles() call.

This has caused the slow tor startup of about 20-30 sec. Without that it is about 5 sec.

sqrrm commented on common/src/main/java/bisq/common/util/Utilities.java:
```diff
@@ -523,4 +525,11 @@ public static int byteArrayToInteger(byte[] bytes) {
     }
 
+    public static String readableFileSize(long size) {
+        if (size <= 0) return "0";
```
Not our usual style, but I quite like this way.

chimp1984 replied:
Was copied from a web resource ;-)

sqrrm commented on core/src/main/java/bisq/core/network/p2p/inventory/GetInventoryRequestHandler.java:
```diff
@@ -0,0 +1,176 @@
+
+    @Override
+    public void onMessage(NetworkEnvelope networkEnvelope, Connection connection) {
+        if (networkEnvelope instanceof GetInventoryRequest) {
```
As always, prefer to return early instead of huge blocks.

sqrrm commented on core/src/main/java/bisq/core/network/p2p/inventory/GetInventoryRequestHandler.java:
```diff
@@ -0,0 +1,176 @@
+            // Data
+            GetInventoryRequest getInventoryRequest = (GetInventoryRequest) networkEnvelope;
+            Map<InventoryItem, Integer> dataObjects = new HashMap<>();
+            p2PDataStorage.getMapForDataResponse(getInventoryRequest.getVersion()).values().stream()
+                    .map(e -> e.getClass().getSimpleName())
+                    .forEach(className -> {
+                        Optional<InventoryItem> optionalEnum = Enums.getIfPresent(InventoryItem.class, className);
+                        if (optionalEnum.isPresent()) {
+                            InventoryItem key = optionalEnum.get();
+                            dataObjects.putIfAbsent(key, 0);
+                            int prev = dataObjects.get(key);
+                            dataObjects.put(key, prev + 1);
+                        }
+                    });
+            p2PDataStorage.getMap().values().stream()
+                    .map(ProtectedStorageEntry::getProtectedStoragePayload)
+                    .filter(Objects::nonNull)
+                    .map(e -> e.getClass().getSimpleName())
+                    .forEach(className -> {
+                        Optional<InventoryItem> optionalEnum = Enums.getIfPresent(InventoryItem.class, className);
+                        if (optionalEnum.isPresent()) {
+                            InventoryItem key = optionalEnum.get();
+                            dataObjects.putIfAbsent(key, 0);
+                            int prev = dataObjects.get(key);
+                            dataObjects.put(key, prev + 1);
+                        }
+                    });
```
```suggestion
            p2PDataStorage.getMapForDataResponse(getInventoryRequest.getVersion()).values().stream()
                    .map(e -> e.getClass().getSimpleName())
                    .forEach(className -> addItem(dataObjects, className));
            p2PDataStorage.getMap().values().stream()
                    .map(ProtectedStorageEntry::getProtectedStoragePayload)
                    .filter(Objects::nonNull)
                    .map(e -> e.getClass().getSimpleName())
                    .forEach(className -> addItem(dataObjects, className));
```

With `addItem()`
```
    private void addItem(Map<InventoryItem, Integer> dataObjects, String className) {
        Optional<InventoryItem> optionalEnum = Enums.getIfPresent(InventoryItem.class, className);
        if (optionalEnum.isPresent()) {
            InventoryItem key = optionalEnum.get();
            dataObjects.putIfAbsent(key, 0);
            int prev = dataObjects.get(key);
            dataObjects.put(key, prev + 1);
        }
    }

```

chimp1984 committed:
Improve logs

sqrrm commented on inventory/src/main/java/bisq/inventory/InventoryMonitorMain.java:
```diff
@@ -0,0 +1,127 @@
+        Res.setup(); // Used for some formatting in the webserver
+
+        // We do not set any capabilities as we don't want to receive any network data beside our response.
+        // We also do not use capabilities for the request/response messages as we only connect to seeds nodes and
```
This comment is lacking a bit when it comes to the punch line.

sqrrm reviewed (commented):
Looks good. Some comments that don't really have to be tended to.

sqrrm reviewed (approved):
utACK

@chimp1984 Comments can be fixed elsewhere if you feel like it, I want to merge this.
